staging/src/k8s.io/kubectl/pkg/cmd/cp/filespec.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cp
+
+import (
+	"path"
+	"path/filepath"
+	"strings"
+)
+
+type fileSpec struct {
+	PodName      string
+	PodNamespace string
+	File         pathSpec
+}
+
+type pathSpec interface {
+	String() string
+	Dir() pathSpec
+	Base() pathSpec
+	Clean() pathSpec
+	Join(pathSpec) pathSpec
+	Glob() (matches []string, err error)
+	Prefix() pathSpec
+	StripShortcuts() pathSpec
+	// IsRelative returns true if dest is pointing outside the base directory, false otherwise.
+	IsRelative(pathSpec) bool
+}
+
+// localPath represents a client-native path, which will differ based
+// on the client OS, its methods will use path/filepath package which
+// is OS dependant
+type localPath struct {
+	file string
+}
+
+func newLocalPath(fileName string) pathSpec {
+	file := stripTrailingSlash(fileName)
+	return localPath{file: file}
+}
+
+func (p localPath) String() string {
+	return p.file
+}
+
+func (p localPath) Dir() pathSpec {
+	return newLocalPath(filepath.Dir(p.file))
+}
+
+func (p localPath) Base() pathSpec {
+	return newLocalPath(filepath.Base(p.file))
+}
+
+func (p localPath) Clean() pathSpec {
+	return newLocalPath(filepath.Clean(p.file))
+}
+
+func (p localPath) Join(elem pathSpec) pathSpec {
+	return newLocalPath(filepath.Join(p.file, elem.String()))
+}
+
+func (p localPath) Glob() (matches []string, err error) {
+	return filepath.Glob(p.file)
+}
+
+func (p localPath) Prefix() pathSpec {
+	return newLocalPath(stripLeadingSlash(p.file))
+}
+
+func (p localPath) StripShortcuts() pathSpec {
+	p = (p.Clean()).(localPath)
+	return newLocalPath(stripPathShortcuts(p.file))
+}
+
+func (p localPath) IsRelative(dest pathSpec) bool {
```

Comment:
nit: consider asserting that pathSpec is local. It's probably an error to ask if a remote path is relative to a local path.

Reply:
on second thought, maybe this (and `Join`) should take the typed (local/remote) path as the argument? That doesn't work with the interface, but it might make sense to just remove these from the interface, and put the onus on the caller to ensure that the right method is being called.